### PR TITLE
external: round robin to use multiple GCS clients

### LINKED ISSUE #50413: Building connection costs lots of time when read multiple small files in Google Cloud Storage
## Enhancement

as title. After we use separate connection for each read request in https://github.com/pingcap/tidb/issues/49340 , I see in the CPU profile lots of time is spent on TLS handshake when creating connection

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #50413

Problem Summary:

### What changed and how does it work?

I noticed that when we read ~1000 small files from GCS, the time of creating the connection is not negligible.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

```
master:
=== RUN   TestSpeedReadManyFiles
    gcs_test.go:541: read 10 small files cost 184.228475ms
    gcs_test.go:541: read 100 small files cost 919.65221ms
    gcs_test.go:541: read 1000 small files cost 10.932206014s
    gcs_test.go:571: read 30 large files cost 4.31413421s
--- PASS: TestSpeedReadManyFiles (50.51s)

delete high throughput client:
=== RUN   TestSpeedReadManyFiles
    gcs_test.go:541: read 10 small files cost 162.914353ms
    gcs_test.go:541: read 100 small files cost 318.649369ms
    gcs_test.go:541: read 1000 small files cost 865.962228ms
    gcs_test.go:571: read 30 large files cost 15.210720205s
--- PASS: TestSpeedReadManyFiles (61.11s)

roundrobin:
=== RUN   TestSpeedReadManyFiles
    gcs_test.go:541: read 10 small files cost 169.609487ms
    gcs_test.go:541: read 100 small files cost 295.653485ms
    gcs_test.go:541: read 1000 small files cost 846.523461ms
    gcs_test.go:571: read 30 large files cost 4.289098557s
--- PASS: TestSpeedReadManyFiles (33.66s)
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/storage/gcs.go [hunk 1]
```diff
@@ -97,14 +98,23 @@ func (options *GCSBackendOptions) parseFromFlags(flags *pflag.FlagSet) error {
 // GCSStorage defines some standard operations for BR/Lightning on the GCS storage.
 // It implements the `ExternalStorage` interface.
 type GCSStorage struct {
-	gcs    *backuppb.GCS
-	bucket *storage.BucketHandle
-	cli    *storage.Client
+	gcs       *backuppb.GCS
+	idx       *atomic.Int64
+	clientCnt int64
+	buckets   []*storage.BucketHandle
```
Comment (D3Hunter) [inline br/pkg/storage/gcs.go:104]:
```suggestion
	handles   []*storage.BucketHandle
```

File: br/pkg/storage/gcs.go [hunk 2]
```diff
@@ -314,6 +324,8 @@ func (s *GCSStorage) Rename(ctx context.Context, oldFileName, newFileName string
 // used in tests
 var mustReportCredErr = false
 
+const gcsClientCnt = 16
```
Comment (D3Hunter) [inline br/pkg/storage/gcs.go:327]: make it configurable?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/storage/gcs.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/storage/gcs.go:104]: it's bucket handle and GCS package also have other handles like object handles, so I think the current name is good

File: br/pkg/storage/gcs.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/storage/gcs.go:327]: I'll leave it for future performance analysis.

File: br/pkg/storage/gcs.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/storage/gcs.go:104]: it reads like we are using multiple `gcs buckets`, but is multiple handles of same bucket actually

**Comment by lance6716** [comment]
/hold

in a 1T test meet this problem https://github.com/googleapis/google-cloud-go/issues/7440

**Review by wjhuang2016 — APPROVED** [review]

**Comment by lance6716** [comment]
with new retry strategy, this PR will cause http2 package consume more memory and 1T IMPORT INTO test will OOM. Will refine the memory consumption part later.

image

**Review by YuJuncen — APPROVED** [review]

**Comment by lance6716** [comment]
/unhold

no longer OOM, and the time duration of 1T test with 3 tidb 3 tikv is 1h40min -> 1h26min

**Pull request merged**